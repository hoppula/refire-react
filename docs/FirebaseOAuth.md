# FirebaseOAuth

> Wrapper component for easy Firebase OAuth login

Available providers are `facebook`, `google`, `twitter` and `github`.

Available authentication flows are `popup` and `redirect`.

You can also pass `scopes` prop to request extra information in OAuth authentication dialog, e.g. obtaining birthday with Facebook authentication you'd pass `['user_birthday']` as prop.

You can also pass `onClick` and `onError` callback functions as props.

Renders given child component.

```js
import { FirebaseOAuth } from 'refire-react'
const LoginWithGoogle = () => {
  return (
    <FirebaseOAuth provider="google" flow="popup">
      <Button>Login with Google</Button>
    </FirebaseOAuth>
  )
}
```
