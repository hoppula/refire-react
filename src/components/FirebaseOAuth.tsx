import * as PropTypes from "prop-types"
import * as React from "react"
import { Component } from "react"
import * as ReactRedux from "react-redux"
import { connect } from "react-redux"
import { firebaseActions } from "refire"
const { oAuthLogin } = firebaseActions
const validProviders = ["facebook", "google", "twitter", "github"]
const validFlows = ["popup", "redirect"]

export interface OAuthProps {
  readonly children?: any,
  readonly dispatch?: ReactRedux.Dispatch<{}>
  readonly flow: "popup" | "redirect",
  readonly onClick?: (event?: React.FormEvent<HTMLElement>) => void,
  readonly onError?: (error: OAuthError) => void,
  readonly provider: "facebook" | "google" | "twitter" | "github",
  readonly scopes?: string[]
}

export interface OAuthError { message: string }

export class FirebaseOAuth extends Component<OAuthProps & ReactRedux.DispatchProp<any>, {}> {

  static propTypes = {
    flow: PropTypes.oneOf(validFlows),
    onClick: PropTypes.func,
    onError: PropTypes.func,
    provider: PropTypes.oneOf(validProviders),
    scopes: PropTypes.array
  }

  authenticate = () => {
    const flow = this.props.flow || "popup"
    this.props.dispatch(
      oAuthLogin(flow, this.props.provider, this.props.scopes)
    ).catch((error: OAuthError) => {
      if (typeof this.props.onError === "function") {
        this.props.onError(error)
      }
    })
    if (typeof this.props.onClick === "function") {
      this.props.onClick()
    }
  }

  render() {
    return React.Children.only(
      React.cloneElement(this.props.children, {
        onClick: this.authenticate
      })
    )
  }

}

export default connect()(FirebaseOAuth)
