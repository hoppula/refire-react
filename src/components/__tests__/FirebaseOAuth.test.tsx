import * as React from "react"
import * as renderer from "react-test-renderer"

import { FirebaseOAuth } from "../FirebaseOAuth"

it("FirebaseOAuth renders given children and calls onClick on click", () => {
  const loginFn = jest.fn()
  const dispatch = () => {
    return new Promise(() => {})
  }

  let component = renderer.create(
    <FirebaseOAuth
      flow="popup"
      provider="facebook"
      dispatch={ dispatch }
      onClick={ loginFn }
    >
      <button>Login with Facebook</button>
    </FirebaseOAuth>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onClick()
  expect(loginFn.mock.calls.length).toBe(1)
})
