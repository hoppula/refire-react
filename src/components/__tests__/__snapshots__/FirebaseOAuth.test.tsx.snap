// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseOAuth renders given children and calls onClick on click 1`] = `
<button
  onClick={[Function]}
>
  Login with Facebook
</button>
`;
